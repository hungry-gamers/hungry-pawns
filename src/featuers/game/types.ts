export type PawnSize = 'small' | 'medium' | 'big'

export interface Pawn {
  size: PawnSize
  playerId: string
}

export type Move = {
  size: PawnSize
  mode: 'pawn' | 'shield'
}

export type Location = { rowIndex: number; columnIndex: number }

export type ApplyShieldPayload = Location

export type SpecialMoveName = 'shield'

export type Sequence = `${string}/${number}-${number}/${number}-${number}/${number}-${number}`

export type Cell = { pawn: Pawn | null; shield: { activeInTurn: number; appliedBy: string } }
export type Board = Cell[][]
export type Player = {
  pawns: Record<PawnSize, number>
  capturedPawnsCounter: number
  arePawnsLocked: boolean
  specialMoves: SpecialMoveName[]
}

export type PlayerPayload = { id: string; pawns: Player['pawns'] }

export type PutPawnPayload = { pawnSize: PawnSize } & Location

export type Game = {
  status: 'not-initialized' | 'pregame' | 'in-progress' | 'finished'
  board: Board
  currentPlayerId: string
  turns: Record<number, { playerId: string; move: PutPawnPayload | ApplyShieldPayload }>
  winner: string | undefined
  allowedPawns: PawnSize[]
  sequences: Sequence[]
}
