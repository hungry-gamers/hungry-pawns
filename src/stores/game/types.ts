export type PawnSize = 'small' | 'medium' | 'big'

export interface Pawn {
  size: PawnSize
  playerId: string
}

export type Cell = Pawn | null
export type Board = (Cell | null)[][]
export type Player = {
  id: string
  pawns: {
    small: number
    medium: number
    big: number
  }
}

export type PutPawnPayload = { pawnSize: PawnSize; rowIndex: number; columnIndex: number }

export type Game = {
  status: 'not-initialized' | 'pregame' | 'in-progress' | 'finished'
  board: Board
  currentPlayerId: string
  pawns: Record<string, Record<PawnSize, number>>
  pawnsLockedBy: string[]
  turns: Record<number, { playerId: string; move: PutPawnPayload }>
  potentialWinner: string | undefined
  eatenPawnsCounter: Record<string, number>
}

type WinCondition = 'captured-line' | 'pawns-eaten' | 'balance' | 'dominance'

export type WinData = { winnerId: string; winCondition: WinCondition } | { winnerId: undefined }
