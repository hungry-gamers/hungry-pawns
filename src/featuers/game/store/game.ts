import { reactive } from 'vue'
import { defineStore } from 'pinia'
import * as GameT from '@/featuers/game/types.ts'
import * as PlayersT from '@/featuers/players/types'
import * as GameService from '@/featuers/game/services/game.service.ts'
import { usePlayersStore } from '@/featuers/players/store/players.ts'
import { SpecialEffectsSequences } from '@/featuers/game/services/game.service.ts'

export const useGameStore = defineStore('game', () => {
  const playersStore = usePlayersStore()

  const state = reactive<GameT.Game>({
    status: 'not-initialized',
    board: GameService.createBoard(),
    currentPlayerId: '',
    turns: {},
    winner: undefined,
    allowedPawns: ['small'],
    sequences: [],
  })

  const initiateGame = (players: PlayersT.PlayerPayload[]) => {
    state.status = 'pregame'
    state.board = GameService.createBoard()
    state.currentPlayerId = players[0].id
    state.allowedPawns = ['small']
    state.sequences = []
    state.winner = undefined

    playersStore.createPlayers(players)
  }

  const lockPawns = (playerId: string, pawns: Record<GameT.PawnSize, number>) => {
    if (playersStore.lockPawns(playerId, pawns)) {
      state.status = 'in-progress'
    }
  }

  const getLastTurn = () => {
    return Math.max(...Object.keys(state.turns).map(Number), 0)
  }

  const getCurrentTurn = () => {
    return getLastTurn() + 1
  }

  const manageAllowedPawns = () => {
    const lastTurn = getLastTurn()
    const rejectConditions = [
      lastTurn < GameService.UNLOCK_MEDIUM_PAWNS_TURN,
      state.allowedPawns.length === 2 && lastTurn < GameService.UNLOCK_BIG_PAWNS_TURN,
      lastTurn > GameService.UNLOCK_BIG_PAWNS_TURN,
    ]

    if (rejectConditions.some(Boolean)) return
    if (lastTurn === GameService.UNLOCK_MEDIUM_PAWNS_TURN) state.allowedPawns.push('medium')
    if (lastTurn === GameService.UNLOCK_BIG_PAWNS_TURN) state.allowedPawns.push('big')
  }

  const nextTurn = (payload?: GameT.PutPawnPayload | GameT.ApplyShieldPayload) => {
    playersStore.decreaseParalyzeProtection(state.currentPlayerId)
    state.turns[getLastTurn() + 1] = { playerId: state.currentPlayerId, move: payload || 'skip' }
    state.currentPlayerId = playersStore
      .getPlayers()
      .find((player) => player !== state.currentPlayerId)!

    manageAllowedPawns()
  }

  const checkInstantWin = () => {
    const CAPTURED_PAWNS_FOR_INSTANT_WIN = 5
    const player = playersStore.getPlayer(state.currentPlayerId)

    return player.capturedPawnsCounter === CAPTURED_PAWNS_FOR_INSTANT_WIN
  }

  const isLineCaptureWin = (): string | undefined => {
    const sequences = GameService.findWinningLine(state.board)

    if (state.sequences.length > 0) {
      const sequence = state.sequences.find((cell) => sequences.includes(cell))

      if (sequence) {
        const [playerId] = sequence.split('/')
        return playerId
      }
    }

    state.sequences = sequences

    return undefined
  }

  const skipTurn = () => {
    const result = playersStore.updateSkippedTurnsCount(state.currentPlayerId)
    const opponentId = playersStore.getOpponentId(state.currentPlayerId)

    const penalties: Record<number, () => void> = {
      1: () => playersStore.manipulatePawnAmount(opponentId, 'small', 1),
      2: () => playersStore.manipulatePawnAmount(opponentId, 'medium', 1),
      3: () => playersStore.manipulatePawnAmount(opponentId, 'big', 1),
      4: () => playersStore.addSpecialMove(opponentId, 'shield'),
      5: () => playersStore.addSpecialMove(opponentId, 'drop'),
    }

    penalties[result]()

    playersStore.removeParalyzeEffect(state.currentPlayerId)

    nextTurn()
  }

  const checkWinner = (payload: GameT.MovePayload) => {
    const isInstantWin = checkInstantWin()
    state.winner = isInstantWin ? state.currentPlayerId : isLineCaptureWin()

    if (!isInstantWin) nextTurn(payload)
    if (state.currentPlayerId === state.winner) {
      state.status = 'finished'
    }
  }

  const isCellProtected = ({ rowIndex, columnIndex }: GameT.Location) => {
    return state.board[rowIndex][columnIndex].shield.activeInTurn === getCurrentTurn()
  }

  const isPutPawnAllowed = (payload: GameT.PutPawnPayload) => {
    const player = playersStore.getPlayer(state.currentPlayerId)
    const { rowIndex, columnIndex, pawnSize } = payload
    const isPawnAvailable = player.pawns[payload.pawnSize] > 0
    const isPawnSizeAllowed = state.allowedPawns.includes(pawnSize)

    return (
      isPawnAvailable &&
      isPawnSizeAllowed &&
      !isCellProtected({ rowIndex, columnIndex }) &&
      !playersStore.getPlayer(state.currentPlayerId).paralyzed.isActive
    )
  }

  const isDropPawnAllowed = (payload: GameT.DropOpponentPawnPayload) => {
    const { rowIndex, columnIndex } = payload
    const isPlayerPawn = state.board[rowIndex][columnIndex].pawn?.playerId === state.currentPlayerId

    return (
      !isCellProtected({ rowIndex, columnIndex }) &&
      !isPlayerPawn &&
      !playersStore.getPlayer(state.currentPlayerId).paralyzed.isActive
    )
  }

  const dropOpponentPawn = (payload: GameT.MovePayload) => {
    if (!isDropPawnAllowed(payload)) return
    if (playersStore.useSpecialMove(state.currentPlayerId, 'drop') === 'not-allowed') return

    const { rowIndex, columnIndex } = payload
    state.board[rowIndex][columnIndex].pawn = null

    checkWinner(payload)
  }

  const paralyzePlayer = () => {
    const opponentId = playersStore.getOpponentId(state.currentPlayerId)
    const result = SpecialEffectsSequences.paralyze.every(([row, column]) => {
      return state.board[row][column].pawn?.playerId === state.currentPlayerId
    })

    if (!result || !playersStore.canPlayerBeParalyzed(opponentId)) return

    playersStore.paralyzePlayer(playersStore.getOpponentId(state.currentPlayerId))
  }

  const putPawn = (payload: GameT.PutPawnPayload) => {
    const { rowIndex, columnIndex, pawnSize } = payload
    const cellPawn = state.board[rowIndex][columnIndex].pawn

    if (!isPutPawnAllowed(payload) || state.status !== 'in-progress') return

    if (cellPawn) {
      if (!GameService.isPawnBigger(pawnSize, cellPawn.size)) return
      if (cellPawn.playerId !== state.currentPlayerId) {
        playersStore.updatePlayerCapturedPawns(state.currentPlayerId)
      }
    }

    state.board[rowIndex][columnIndex].pawn = {
      size: pawnSize,
      playerId: state.currentPlayerId,
    }
    playersStore.manipulatePawnAmount(state.currentPlayerId, pawnSize, -1)

    paralyzePlayer()
    checkWinner(payload)
  }

  const applyShield = (payload: GameT.ApplyShieldPayload) => {
    if (playersStore.getPlayer(state.currentPlayerId).paralyzed.isActive) return
    if (playersStore.useSpecialMove(state.currentPlayerId, 'shield') === 'not-allowed') return
    const { rowIndex, columnIndex } = payload

    state.board[rowIndex][columnIndex].shield.appliedBy = state.currentPlayerId
    state.board[rowIndex][columnIndex].shield.activeInTurn = getLastTurn() + 2

    checkWinner(payload)
  }

  return {
    state,
    initiateGame,
    putPawn,
    lockPawns,
    applyShield,
    getCurrentTurn,
    isLineCaptureWin,
    dropOpponentPawn,
    skipTurn,
  }
})
