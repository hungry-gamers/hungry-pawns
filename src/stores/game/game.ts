import { reactive } from 'vue'
import { defineStore } from 'pinia'
import type { Game, PutPawnPayload, Player, PawnSize, Cell } from '@/stores/game/types.ts'

export const useGameStore = defineStore('game', () => {
  const state = reactive<Game>({
    status: 'not-initialized',
    board: Array.from({ length: 3 }, () => Array(3).fill(null)),
    currentPlayerId: '',
    pawns: {},
    pawnsLockedBy: [],
    turns: {},
    potentialWinner: undefined,
  })

  const initiateGame = (players: Player[]) => {
    state.status = 'pregame'
    state.board = Array.from({ length: 3 }, () => Array(3).fill(null))
    state.currentPlayerId = players[0].id

    players.forEach((player) => {
      state.pawns[player.id] = { ...player.pawns }
    })
  }

  const getPlayers = (): string[] => {
    return Object.keys(state.pawns ?? {})
  }

  const lockPawns = (playerId: string, pawns: Record<PawnSize, number>) => {
    if (state.pawnsLockedBy.includes(playerId)) return
    state.pawnsLockedBy.push(playerId)
    state.pawns[playerId].small = pawns.small
    state.pawns[playerId].medium = pawns.medium
    state.pawns[playerId].big = pawns.big
    const [player1Id, player2Id] = getPlayers()

    if (state.pawnsLockedBy.includes(player1Id) && state.pawnsLockedBy.includes(player2Id)) {
      state.status = 'in-progress'
    }
  }

  const isBigger = (size: PawnSize, pawnOnBoardSize: PawnSize) => {
    const order: PawnSize[] = ['small', 'medium', 'big']
    return order.indexOf(size) > order.indexOf(pawnOnBoardSize)
  }

  const checkWinner = (): string | undefined => {
    const board = state.board

    const isWinningLine = (line: Cell[]) => {
      if (line.some((cell) => !cell)) return false
      const playerId = line[0]?.playerId
      return line.every((cell) => cell?.playerId === playerId)
    }

    for (let i = 0; i < 3; i++) {
      if (isWinningLine([board[i][0], board[i][1], board[i][2]])) return board[i][0]?.playerId
      if (isWinningLine([board[0][i], board[1][i], board[2][i]])) return board[0][i]?.playerId
    }

    if (isWinningLine([board[0][0], board[1][1], board[2][2]])) return board[0][0]?.playerId
    if (isWinningLine([board[0][2], board[1][1], board[2][0]])) return board[0][2]?.playerId

    return undefined
  }

  const putPawn = (payload: PutPawnPayload) => {
    if (!state.pawns[state.currentPlayerId][payload.pawnSize]) return
    const cellPawn = state.board[payload.rowIndex][payload.columnIndex]

    if (cellPawn && !isBigger(payload.pawnSize, cellPawn.size)) return

    const players = Object.keys(state.pawns)
    state.board[payload.rowIndex][payload.columnIndex] = {
      size: payload.pawnSize,
      playerId: state.currentPlayerId,
    }
    state.pawns[state.currentPlayerId][payload.pawnSize]--
    const lastTurn = Object.keys(state.turns).reverse()[0] ?? 0
    state.turns[+lastTurn + 1] = { playerId: state.currentPlayerId, move: payload }
    state.potentialWinner = checkWinner()
    state.currentPlayerId = players.find((player) => player !== state.currentPlayerId)!

    if (state.currentPlayerId === state.potentialWinner) {
      state.status = 'finished'
      return state.potentialWinner
    }
  }

  return {
    state,
    initiateGame,
    putPawn,
    getPlayers,
    lockPawns,
    checkWinner,
  }
})
