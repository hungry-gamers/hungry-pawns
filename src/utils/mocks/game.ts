import type { Player } from '@/stores/game/types.ts'

export const players: Player[] = [
  { id: '1', pawns: { small: 3, medium: 3, big: 3 } },
  { id: '2', pawns: { small: 3, medium: 3, big: 3 } },
]
