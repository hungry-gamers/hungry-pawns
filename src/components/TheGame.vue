<script setup lang="ts">
import { useGameStore } from '@/stores/game/game.ts'
import type { PawnSize } from '@/stores/game/types.ts'

const props = defineProps<{ pawnSize: PawnSize }>()

const { putPawn, state } = useGameStore()

const onCellClick = (rowIndex: number, columnIndex: number) => {
  putPawn({
    pawn: { size: props.pawnSize, player: state.currentPlayerId },
    rowIndex,
    columnIndex,
  })
}
</script>

<template>
  <div id="board">
    <div v-for="(row, rowIndex) in state.board" :key="rowIndex" class="row">
      <div
        v-for="(cell, colIndex) in row"
        :key="colIndex"
        :data-test-id="`cell-${rowIndex}-${colIndex}`"
        class="cell"
        @click="onCellClick(rowIndex, colIndex)"
      >
        <span v-if="cell">{{ cell.size }}</span>
      </div>
    </div>
  </div>
</template>

<style scoped>
#board {
  display: grid;
  grid-template-rows: repeat(3, 200px);
  gap: 5px;
  width: 610px;
}

.row {
  display: grid;
  grid-template-columns: repeat(3, 200px);
  gap: 5px;
}

.cell {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 200px;
  height: 200px;
  background-color: white;
  border: 1px solid black;
  cursor: pointer;
}

.cell:hover {
  background-color: #f0f0f0;
}
</style>
