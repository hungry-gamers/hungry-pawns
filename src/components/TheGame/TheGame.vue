<script setup lang="ts">
import { useGameStore } from '@/stores/game/game.ts'
import type { PawnSize } from '@/stores/game/types.ts'
import { computed } from 'vue'

const props = defineProps<{ pawnSize: PawnSize }>()

const { putPawn, state, getPlayers } = useGameStore()

const players = computed(getPlayers)

const onCellClick = (rowIndex: number, columnIndex: number) => {
  putPawn({
    pawnSize: props.pawnSize,
    rowIndex,
    columnIndex,
  })
}
</script>

<template>
  <div id="board">
    <div v-for="(row, rowIndex) in state.board" :key="rowIndex" class="row">
      <div
        v-for="(cell, colIndex) in row"
        :key="colIndex"
        :data-test-id="`cell-${rowIndex}-${colIndex}`"
        class="cell"
        :class="{
          'player-1': cell?.playerId === players[0],
          'player-2': cell?.playerId === players[1],
        }"
        @click="onCellClick(rowIndex, colIndex)"
      >
        <span v-if="cell">{{ cell.size }}</span>
      </div>
    </div>
  </div>
</template>

<style scoped>
#board {
  display: grid;
  grid-template-rows: repeat(3, 200px);
  gap: 5px;
  width: 610px;
}

.row {
  display: grid;
  grid-template-columns: repeat(3, 200px);
  gap: 5px;
}

.cell {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 200px;
  height: 200px;
  background-color: white;
  border: 2px solid black;
  cursor: pointer;
}

.cell.player-1 {
  border: 2px solid darkorange;
}

.cell.player-2 {
  border: 2px solid dodgerblue;
}

.cell:hover {
  background: #d1d1d1;
}
</style>
