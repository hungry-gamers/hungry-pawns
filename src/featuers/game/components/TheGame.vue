<script setup lang="ts">
import { useGameStore } from '@/featuers/game/store/game.ts'
import type { Move } from '@/featuers/game/types.ts'
import { computed } from 'vue'
import CellShield from '@/featuers/game/components/CellShield.vue'
import ThePawn from '@/featuers/game/components/ThePawn.vue'
import { usePlayersStore } from '@/featuers/players/store/players.ts'

const props = defineProps<{ move: Move }>()

const { putPawn, state, applyShield, getCurrentTurn } = useGameStore()
const { getPlayers } = usePlayersStore()

const players = computed(getPlayers)

const onCellClick = (rowIndex: number, columnIndex: number) => {
  if (props.move.mode === 'pawn') {
    putPawn({
      pawnSize: props.move.size,
      rowIndex,
      columnIndex,
    })
  } else {
    applyShield({ rowIndex, columnIndex })
  }
}
</script>

<template>
  <div id="board">
    <div v-for="(row, rowIndex) in state.board" :key="rowIndex" class="row">
      <div
        v-for="(cell, colIndex) in row"
        :key="colIndex"
        :data-test-id="`cell-${rowIndex}-${colIndex}`"
        class="cell"
        :class="{
          'player-1': cell.pawn?.playerId === players[0],
          'player-2': cell.pawn?.playerId === players[1],
        }"
        @click="onCellClick(rowIndex, colIndex)"
      >
        <CellShield
          v-if="cell.shield.activeInTurn === getCurrentTurn()"
          :owner-id="cell.shield.appliedBy"
        />

        <ThePawn
          v-if="cell.pawn"
          :size="cell.pawn.size"
          :color="cell.pawn?.playerId === players[0] ? 'darkorange' : 'dodgerblue'"
        />
      </div>
    </div>
  </div>
</template>

<style scoped>
.cell:hover > :deep(svg .happy) {
  visibility: hidden;
}

.cell:hover > :deep(svg .scared) {
  visibility: visible;
}

#board {
  display: grid;
  grid-template-rows: repeat(3, 200px);
  gap: 5px;
  width: 610px;
}

.row {
  display: grid;
  grid-template-columns: repeat(3, 200px);
  gap: 5px;
}

.cell {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 200px;
  height: 200px;
  background-color: white;
  border: 2px solid black;
  cursor: pointer;
}

.cell.player-1 {
  border: 2px solid darkorange;
}

.cell.player-2 {
  border: 2px solid dodgerblue;
}

.cell:hover {
  background: #d1d1d1;
}
</style>
