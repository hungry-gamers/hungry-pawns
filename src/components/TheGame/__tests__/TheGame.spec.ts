import { describe, it, expect, vi } from 'vitest'
import { createTestingPinia } from '@pinia/testing'

import { mount } from '@vue/test-utils'
import TheGame from '../TheGame.vue'
import { useGameStore } from '../../../stores/game/game'

describe('TheGame', () => {
  const pinia = createTestingPinia({
    createSpy: vi.fn,
  })

  const gameStore = useGameStore()
  gameStore.getPlayers = vi.fn(() => ['1', '2'])

  const wrapper = mount(TheGame, {
    props: { move: { size: 'small', mode: 'pawn' } },
    global: {
      plugins: [pinia],
    },
  })

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should put pawn on the board', () => {
    const cell = wrapper.find('[data-test-id="cell-0-0"]')
    gameStore.state.currentPlayerId = '1'

    cell.trigger('click')

    expect(gameStore.putPawn).toHaveBeenCalledTimes(1)
    expect(gameStore.putPawn).toHaveBeenCalledWith({
      pawnSize: 'small',
      rowIndex: 0,
      columnIndex: 0,
    })
  })
})
