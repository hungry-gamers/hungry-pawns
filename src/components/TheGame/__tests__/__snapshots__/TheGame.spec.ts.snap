// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TheGame > should render correctly 1`] = `
"<div data-v-bc1fe25a="" id="board">
  <div data-v-bc1fe25a="" class="row">
    <div data-v-bc1fe25a="" data-test-id="cell-0-0" class="cell">
      <!--v-if-->
    </div>
    <div data-v-bc1fe25a="" data-test-id="cell-0-1" class="cell">
      <!--v-if-->
    </div>
    <div data-v-bc1fe25a="" data-test-id="cell-0-2" class="cell">
      <!--v-if-->
    </div>
  </div>
  <div data-v-bc1fe25a="" class="row">
    <div data-v-bc1fe25a="" data-test-id="cell-1-0" class="cell">
      <!--v-if-->
    </div>
    <div data-v-bc1fe25a="" data-test-id="cell-1-1" class="cell">
      <!--v-if-->
    </div>
    <div data-v-bc1fe25a="" data-test-id="cell-1-2" class="cell">
      <!--v-if-->
    </div>
  </div>
  <div data-v-bc1fe25a="" class="row">
    <div data-v-bc1fe25a="" data-test-id="cell-2-0" class="cell">
      <!--v-if-->
    </div>
    <div data-v-bc1fe25a="" data-test-id="cell-2-1" class="cell">
      <!--v-if-->
    </div>
    <div data-v-bc1fe25a="" data-test-id="cell-2-2" class="cell">
      <!--v-if-->
    </div>
  </div>
</div>"
`;
